### ISSUE TITLE
New feature concept: counting days that meet user-specified criteria

### ISSUE THREAD

**Issue by mastefan** [bluegreen-labs/daymetr#25]
The code below counts the number of days that meet a user-specified condition, and returns a spatially-explicit result as either a raster object or file. This tool is able to extract useful indices like Growing Degree Days.

I put this together to extract GDD, freezing degree days, and as a data exploration tool.

This code is currently reliant on sapply. I am working on making it more efficient using raster::overlay and raster::calc.

@khufkens do you think that this would fit within the package?

``` 
#' @param data string containing the name of a file
#'  containing the daily Daymet data to be evaluated
#' @param startday numeric day of year at which counting should begin. 
#' Default = 0 (January 1)
#' @param endday numeric day of year at which counting should end. 
#' Default = 365 (December 31)
#' @param criteria logical expression (e.g. ">") to be evaluated
#' @param value numeric the value that the criteria is to be evaluated
#'  against
#' @param internal TRUE / FALSE (if FALSE, write the output to file)
#' using the Daymet file format protocol.
#' 
#' @description This tool counts the number of days in a given time period
#' that meet a given set of criteria. This can be used to extract indices 
#' such as Growing Degree Days (tmin > 0), or days with precipitation 
#' (prcp != 0).

calc_nd <- function(data = NULL,
                    startday = 0,
                    endday = 365,
                    criteria = NULL,
                    value = NULL,
                    internal = FALSE){
  
  # perform input checks
  if(is.null(data) | is.null(criteria) | is.null(value)){
    stop('Please specify data, criteria and value.')
  }
  
  # load data and convert to dataframe
  data <- suppressWarnings(raster::brick(data))
  data.df <- raster::as.data.frame(data,
                                   stringsAsFactors = FALSE)
  
  # select desired dates from data to create working dataset
  data.wrk <- as.data.frame(data.df[ , ,startday:endday],
                            stringsAsFactors = FALSE)
  
  # use boolean indexing in sapply to subset working data
  sel <- sapply(data.wrk,
                criteria,
                value)
  
  # convert df to array
  nr       <- raster::nrow(data)
  nc       <- raster::ncol(data)
  nbnd     <- ncol(sel)
  dim(sel) <- c(nr,nc,nbnd)
  
  # convert array to raster
  crs      <- raster::crs(data)
  ext      <- raster::extent(data)
  sel.ras  <- raster::brick(crs = crs,
                            nrows = nr,
                            ncols = nc,
                            nl = nbnd)
  sel.ras <- raster::setValues(sel.ras,
                               sel)
  
    
  # use SUM to gather the number of days that meet the criteria
  result <- raster::calc(sel.ras,
                         sum,
                         na.rm=TRUE)
  
  # return all data to raster, either as a geotiff or as a local object
  if (internal == FALSE){
    # create output file name
    year <- strsplit(input_file, "_")[[1]][3]
    output_file <- sprintf('nd_%s.tif',year)
    
    # write result to file
    raster::writeRaster(result,
                        output_file,
                        overwrite = TRUE)
  } else {
    # return result
    return(result)
  }
}
```

**Comment by khufkens** [comment]
This would be a nice wrapper indeed!

Could you make it possible to set an output directory. This is mostly for data management if you have a ton of tiles. This also goes for your other contributions. It's always valuable to keep scalability in mind, especially given the tiled layout of the Daymet data.

Also consider raster::subset() to do a selection on the yearly tile data. This saves some lines of code. You can use both indices or the layer names if so desired. Keep in mind that the netcdf queries can return multi year datasets. I'm not sure how to handle these within this context.

```
b <- raster::brick(data)
b <- raster::subset(b, start_date:end_date)
fun = function(x){
...
)
result <- raster::calc(b, fun = fun, na.rm = TRUE)
```

Another tip, readability for variables is so that underscore _ delimited variables read easier. In general I prefer start_date over startdate, startDate (camel case) or start.date. You see this _ syntax pop up more often in R if you browse some packages (especially more recent ones such as tidyverse). I try to stick to this for consistency.

Sorry about being so picky about the code. I hope my pointers help you understand these small optimizations.

**Comment by khufkens** [comment]
Just popped into my head. Might want to add long term summaries, or the possibility for it when dealing with a multiyear ncss file. Would allow you to quickly summarize how # of frost days decrease / increase etc. This might require a separate function.

**Comment by mastefan** [comment]
Thanks for being picky! It helps me to learn. I have updated the code to work with raster:: functions. See below...

I added a way to set the output directory, but haven't tested it yet.

raster::stack lets you specify which bands should be read from the file, so I went with that to set start and end.

startday and endday are now start_day and end_day. You are correct, this is easier to read.

Good idea re. multi-year summaries. I'll do some thinking on that...

Any other thoughts on the code?

```
calc_nd <- function(file = NULL,
                    start_day = 0,
                    end_day = 365,
                    criteria = NULL,
                    value = NULL,
                    internal = FALSE,
                    output_directory = NULL){
  
  # perform input checks
  if(is.null(file) | is.null(criteria) | is.null(value)){
    stop('Please specify file, criteria and value.')
  }
  
  # load desired bands from file
  data <- suppressWarnings(raster::stack(file,
                                         bands = c(start_day:end_day)))

  # use a binary operator to identify pixels that meet the criteria
  sel <- raster::overlay(x = data,
                         fun = function(x) do.call(criteria, list(x, value)))
  
  # use SUM to gather the number of days that meet the criteria
  result <- raster::calc(x = sel,
                         fun = sum,
                         na.rm = TRUE)
  
  # return all data to raster, either as a geotiff or as a local object
  if (internal == FALSE){
    # create output file name
    year <- strsplit(input_file, "_")[[1]][3]
    output_file <- sprintf('nd_%s.tif',year)
    
    # setting output directory if specified
    if (!is.null(output_directory)){
      output_file <- paste0(output_directory, output_file)
    }
    
    # write result to file
    raster::writeRaster(result,
                        output_file,
                        overwrite = TRUE)
  } else {
    # return result
    return(result)
  }
}
```

**Comment by khufkens** [comment]
Implemented in release 1.3.3
